components/style/gecko_selector_impl.rs

```diff
@@ -12,104 +12,85 @@ use stylesheets::Stylesheet;
 #[derive(Debug, Clone, PartialEq, Eq)]
 pub struct GeckoSelectorImpl;
 
+/// NOTE: The boolean field represents whether this element is an anonymous box.
+///
+/// This is just for convenience, instead of recomputing it. Also, note that
+/// Atom is always a static atom, so if space is a concern, we can use the
+/// raw pointer and use the lower bit to represent it without space overhead.
+///
+/// FIXME(emilio): we know all these atoms are static. Patches are starting to
+/// pile up, but a further potential optimisation is generating bindings without
+/// `-no-gen-bitfield-methods` (that was removed to compile on stable, but it no
+/// longer depends on it), and using the raw *mut nsIAtom (properly asserting
+/// we're a static atom).
+///
+/// This should allow us to avoid random FFI overhead when cloning/dropping
+/// pseudos.
+///
+/// Also, we can further optimize PartialEq and hash comparing/hashing only the
+/// atoms.
 #[derive(Clone, Debug, PartialEq, Eq, Hash)]
-pub enum PseudoElement {
-    Before,
-    After,
-
-    Backdrop,
-    FirstLetter,
-    FirstLine,
-    MozSelection,
-    MozFocusInner,
-    MozFocusOuter,
-    MozListBullet,
-    MozListNumber,
-    MozMathAnonymous,
-    MozNumberWrapper,
-    MozNumberText,
-    MozNumberSpinBox,
-    MozNumberSpinUp,
-    MozNumberSpinDown,
-    MozProgressBar,
-    MozRangeTrack,
-    MozRangeProgress,
-    MozRangeThumb,
-    MozMeterBar,
-    MozPlaceholder,
-    MozColorSwatch,
-
-    AnonBox(AnonBoxPseudoElement),
-}
+pub struct PseudoElement(Atom, bool);
```

Comment:
Maybe use struct syntax with named fields? I particular to say what this `bool` is.